### PR TITLE
Solve issues with gitlab .apk files and cloudflare protection

### PR DESCRIPTION
**Pull request by akramer-zibra**
There occur issues if releases refer to external hosted .apk files. In some cases (e.g. Aurora Store) download is not possible because cloudflare protection gives "forbidden" error. This patch will add a "referer" header to the download request to pacify cloudflare protection service. 

Notice: In case of the `Aurora Store` project, you still need a `Gitlab API token`, to retrieve latest .apk release. Obtainium either retrieves .apk data from Gitlab API (token necessary) or from `tags`-page, which has no information about the latest .apk url.

Related to: #1397, #1389, #1384, #1382, #1381, #1380, #1359, #854, #785, #697

Tested with Android 14 in an emulated AVD.

### CODE REVIEW

**Comment by ImranR98** [comment]
Great 👍

Just to confirm - were you still getting blocked without the Referrer header? I haven't been able to test this because I'm not longer getting blocked by Cloudflare either way (not sure if they removed it or just selective about IPs).

**Comment by Disqu3-mirOir-qt** [comment]
hello @ImranR98 fixed side AuroraStore cf: 

https://gitlab.com/AuroraOSS/AuroraStore/-/issues/1052

**Comment by Grishnackh** [comment]
@Disqu3-mirOir-qt Can you please provide information on the mentioned solution to the Cloudflare issue, i.e. any link to a PR or commit? In the linked ticket above the maintainer only states that you fixed the issue, unfortunately not how this has been fixed.

**Comment by akramer-zibra** [comment]
I can confirm the cloudflare fix. The .apk dowload from `auroraoss.com` now works inside Obtainium. Tested with Android 14, latest Obtainium version and a configured Gitlab API token. Think we can close this PR if anyone else can confirm this?

**Comment by DwainZwerg** [comment]
> Think we can close this PR if anyone else can confirm this?
We should integrate the PR anyway. There are certainly other sites that are protected with Cloudflare and with which Obtainium has problems.

**Comment by Grishnackh** [comment]
I agree, @DwainZwerg. This might solve issues for other sources, i.e. #1387.

**Comment by akramer-zibra** [comment]
> There are certainly other sites that are protected with Cloudflare and with which Obtainium has problems.

So you think this solution should be more generic? 

Seems not a bad idea to send always a `user-agent` with http requests 🤔

**Comment by DwainZwerg** [comment]
> Seems nt a bad idea to send always a `user-agent` with http requests.

I think it could be an option to opt in for each app in the app settings.

**Comment by Grishnackh** [comment]
@mjdNjhNJ It's related to personal access tokens, see https://github.com/ImranR98/Obtainium/wiki#gitlab.

**Comment by mjdNjhNJ** [comment]
@Grishnackh Thanks. Confirming that everything works as expected with personal access token (read_api), Aurora Store successfully updated to 4.4.2.

File: lib/app_sources/gitlab.dart [hunk 1]
```diff
@@ -104,6 +105,21 @@ class GitLab extends AppSource {
...
       '$standardUrl/-/releases';
 
+  @override
+  Future<Map<String, String>?> getRequestHeaders(
+      Map<String, dynamic> additionalSettings,
+      {bool forAPKDownload = false}) async {
+    // Change headers to pacify, e.g. cloudflare protection
+    // Related to: (#1397, #1389, #1384, #1382, #1381, #1380, #1359, #854, #785, #697)
+    var headers = <String, String>{};
+    headers[HttpHeaders.refererHeader] = 'https://gitlab.com';
```
Comment (ImranR98) [inline lib/app_sources/gitlab.dart:115]: This should be `'https://${hosts[0]}'` in case someone overrides the host.

**Comment by ImranR98** [comment]
> So you think this solution should be more generic?

@akramer-zibra let me know if you still want the PR to be merged (after the one review comment is addressed) or if you plan on making something more generic. I don't think there's any harm in adding a referrer header by default for all requests (just not sure if that would be enough to bypass CF protection anyways).

**Comment by akramer-zibra** [comment]
Hey @ImranR98 . I changed the host in the referer header as you mentioned. Sure this change makes the code more consistent.

Tested this with `Android 14` on an emulated AVD:
- [x] _When_ I add `Aurora Store` gitlab repositoy _Then_ Obtainium shows `AuroraStore 4.3.5` as latest version _And When_ I click install _Then_ Obtainium installs this version (4.3.5).
- [x]  _Given_ a `Gitlab API token` in the settings _When_ I add `Aurora Store` gitlab repository _Then_ Obtainium shows a Warning prompt "The App source ... but the release package comes from ..." _And When_ I click continue _Then_ Obtainium shows `AuroraStore 4.4.2` as latest version _And When_ I click install _Then_ Obtainium installs this version (4.4.2).

I think this PR is ready to merge. 

As some people mentioned above, there could be a more generic solution to this: for example a On-Off-switch within the app specific options to enable ("opt-in") sending HTTP headers. I would prefer to handle this within a separate PR, because I think there might be a need for further discussion on this. 

Nevertheless the Gitlab scraping seems still having issues when the .apk information on the release-page and tags-page differ. Using the Gitlab API for .apk retrieval solves this problem, but not every Obtainium user might have a Gitlab.com account. So I think there also might be a need for some code enhancements concerning this. I will endeavour to make a PR for this in the near future.

**Pull request merged**